core/exec/src/mill/exec/Execution.scala

```diff
@@ -188,7 +192,19 @@ private[mill] case class Execution(
                   exclusive
                 )
 
-                if (failFast && res.newResults.values.exists(_.asSuccess.isEmpty))
+                // Count new failures - if there are upstream failures, tasks should be skipped, not failed
+                val newFailures = res.newResults.values.count(r => r.asFailing.isDefined)
+
+                if (newFailures > 0) {
```

Comment:
This conditional isn't necessary right? If it's zero then `addAndGet` is a no-op already, and it will never be negative